src/app/api/device.model.spec.ts

```
@@ -0,0 +1,109 @@
+import { DeviceAdapter } from './device.model';
+import { EntityAdapter } from './entity.model';
+import { MaterialIconAdapter } from './material-icon.model';
+import { TestBed } from '@angular/core/testing';
+
+describe('DeviceAdapter', () => {
+  let deviceAdapter: DeviceAdapter;
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      imports: [],
+      providers: [MaterialIconAdapter, EntityAdapter, DeviceAdapter],
+    });
+
+    deviceAdapter = TestBed.inject(DeviceAdapter);
+  });
+
+  it('should adapt device from full JSON', () => {
+    const json =
+      '{\n' +
+      '    "id": 1,\n' +
+      '    "key": "test",\n' +
+      '    "macAddress": "c0:3e:ba:c3:50:0b",\n' +
+      '    "displayName": "test",\n' +
+      '    "icon": {\n' +
+      '        "iconKey": "thermostat",\n' +
+      '        "displayName": "temperature",\n' +
+      '        "name": "THERMOSTAT"\n' +
+      '    },\n' +
+      '    "entities": [\n' +
+      '        {\n' +
+      '            "id": 1,\n' +
+      '            "key": "temp",\n' +
+      '            "displayName": "temperature",\n' +
+      '            "type": "SENSOR",\n' +
+      '            "unitType": "DECIMAL",\n' +
+      '            "unitDisplayName": "C",\n' +
+      '            "icon": {\n' +
+      '                "iconKey": "thermostat",\n' +
+      '                "displayName": "temperature",\n' +
+      '                "name": "THERMOSTAT"\n' +
+      '            }\n' +
+      '        }\n' +
+      '    ]\n' +
+      '}';
+    const deviceObject = JSON.parse(json);
+    const device = deviceAdapter.adapt(deviceObject);
+
+    expect(device).toBeDefined();
+    if (device) {
```

this will silently fail if `device` is defined but of `null` value, in which case it'll just skip all the assertions below. Using `device!.property` would IMO be the better way to handle this.

\+ same applies to other tests